Allow for multiple endpoints and many-to-many mappings of endpoints to sinks

friso opened the pull request:
This is substantial change that allows to define multiple endpoints as event sources and route events through different mappings, using different schema's and sink them into different sinks (multiple HDFS locations, multiple Kafka topics).

Commit:
Start of work to support many-to-many sources/sinks/mappings.

Changes include:
 - A new configuration layout for supporting multiple sources, sinks and mappings.
 - No more enqueuing delay. Queues tend to either be full or empty, and are really only effective at smoothing jitter. Spinning to delay doesn't really help, so we just drop the events immediately instead of blocking.
 - No more session-binning strategy for HDFS. In practice we didn't use this, and it's complicated to do properly when multiple sources (with their own idea of sessions) might be feeding into a sink.

For now, the code only supports the existing single browser-based source and HDFS/Kafka sinks. Configuration validation is also not yet implemented.

Commit:
Fix the default value for automatic page-view events.

friso committed:
Favor ImmutableProperties over getter with defensive copy.

friso committed:
Use deprecation annotation instead of comment.

friso committed:
Do not throw RuntimeException on MappingException, but instead catch mapping exceptions and turn them into useful error messages and leave configuratio invalid.

friso committed:
Extract some methods for readability.

friso committed:
Validators for DivolteConfiguration.

friso committed:
Reordering of methods.

friso committed:
Flatten ClientSideCookieHandler and super class into one class.

Commit:
Merge branch 'master' into dev/many-to-many

# Conflicts:
#	build.gradle

friso committed:
Remove Optional<...> from *Configuration constructors where actual property is not optional. This is to make the constructor args match the reality of the config params, such that we can later use JavaDoc for the documentation. Default values are set using the defaultValue annotation, which is however not interpreted by the deserializer. We will later fix this using a custom deserializer in Jackson.

friso committed:
Merge branch 'dev/many-to-many' of github.com:divolte/divolte-collector into dev/many-to-many

Commit:
Merge branch 'master' into dev/many-to-many

# Conflicts:
#	src/main/java/io/divolte/server/kafka/KafkaFlusher.java
#	src/main/java/io/divolte/server/kafka/KafkaFlushingPool.java
#	src/main/resources/reference.conf

friso committed:
Store source ID and affinity in items on the processing queues.

friso committed:
Resolve conflict on Kafka flushing code.

Commit:
Merge branch 'master' into dev/many-to-many

Commit:
Use .isEmpty() instead of comparing size to 0.

Commit:
During tests, fail early if the test server configuration isn't valid.

Commit:
Start the test server automatically after construction.

All users of this class did this anyway.

Commit:
Avoid empty blocks.

Commit:
Refactor test configurations.

 - Eliminate a bunch of empty configuration files.
 - Change test layout to let tests run against the default configuration with the test server, or a base convenience configuration.

Commit:
Add a test for the /ping endpoint.

Commit:
Ensure the prefix for a browser source has a trailing '/'.

The browser source requires this or the tracking script won't correctly locate the event end-point.

Commit:
Tweak log messages.

No trailing period (.) required.

Commit:
Start supporting multiple browser endpoints.

Commit:
Convert method to be static.

It's invoked by the constructor, and constructors shouldn't normally invoke instance methods because they might not expect the instance to be partially constructed.

Commit:
Merge branch 'master' into dev/many-to-many

Commit:
Minor optimisations during flushing: pre-size the arrays we're streaming into.

Commit:
Nullability annotations and enforcement.

Commit:
Static references are normally UPPER_CASE.

Commit:
Whitespace.

Commit:
No need for either of these tests to reference the Kafka sink.

Commit:
Support for multiple HDFS and Kafka sinks.

Changes include:
 - Sink configurations now need to supply a factory for creating the sink.
 - The optional GeoIP service is now passed around as such. It's a bit weird why it was nullable in some places before.
 - There's now a central point for loading schemas. This is needed for schema inference with sinks and ensures schemas are only loaded once.
 - Some utilities for collecting streams into Guava's immutable collections.
 - Processing pools are now shutdown by the central server, instead of recursively via the mapper.

Commit:
Fix test configurations, now that we elide disabled sinks.

Commit:
Some basic tests for multiple sinks.

Commit:
Mark utility class as final.

Commit:
Remove an unused logger.

Commit:
Fix incompatible-schema detection for shared sinks.

The associated tests have also been updated; the fixture for the previous test was not correct.

Commit:
Remove some deprecated code.

Commit:
Additional logging during startup.

Commit:
Remove superseded todo.

Commit:
Specialise event forwarding for common cases.

Commit:
Instantiate all mappings during startup, not just one of them.

Commit:
Support many-to-many mappings between sources and mappings.

krisgeus commented on build.gradle:
```diff
@@ -116,6 +116,7 @@ dependencies {
     compile group: 'com.jayway.jsonpath', name: 'json-path', version: '2.1.0'
     compile group: 'com.fasterxml.jackson.core', name:'jackson-databind', version: '2.6.4'
     compile group: 'com.fasterxml.jackson.datatype', name:'jackson-datatype-jdk8', version: '2.6.4'
+    compile group: 'com.fasterxml.jackson.datatype', name:'jackson-datatype-guava', version: '2.6.4'
```
Might want to extract the version number since it's used multiple times related to jackson

krisgeus commented on src/main/java/io/divolte/server/ClientSideCookieEventHandler.java:
```diff
@@ -80,6 +97,68 @@ protected void logEvent(final HttpServerExchange exchange) {
+                .put(Headers.ETAG, SENTINEL_ETAG_VALUE)
+                .put(Headers.CACHE_CONTROL, "private, no-cache, proxy-revalidate")
+                .put(Headers.PRAGMA, "no-cache")
+                .put(Headers.EXPIRES, "Fri, 14 Apr 1995 11:30:00 GMT");
```
Who's birthday is this ;-)

asnare replied:
I have no idea where this date came from. :)

krisgeus commented on src/main/java/io/divolte/server/IncomingRequestProcessor.java:
```diff
@@ -65,42 +56,29 @@
 
-        memory = new ShortTermDuplicateMemory(vc.configuration().incomingRequestProcessor.duplicateMemorySize);
-        keepDuplicates = !vc.configuration().incomingRequestProcessor.discardDuplicates;
+        memory = new ShortTermDuplicateMemory(vc.configuration().global.mapper.duplicateMemorySize);
```
Is it possible to use something like vc.configuration().getGlobalMappingConfiguration().duplicateMemorySize or vc.globalConfiguration().duplicateMemorySize.. If the global.mapper namespace changes you only need 1 place to change it.

krisgeus commented on src/main/java/io/divolte/server/Server.java:
```diff
@@ -135,7 +204,9 @@ public void shutdown() {
+        mappingProcessors.values().forEach(ProcessingPool::stop);
+        sinks.values().forEach(ProcessingPool::stop);
 
         logger.info("Closing HDFS filesystem connection.");
```
Still needed here? Isn't the HDFS specific sink responsible for closing

friso replied:
The HDFS client registers a JVM shutdown hook, which we disable again in the HDFS configuration somewhere, because we cannot control the ordering of shutdown hooks, meaning HDFS may close before we flush on shutdown. Hence, we close all HDFS stuff explicitly in our own shutdown sequence.

asnare replied:
@krisgeus The sinks close their files, but not the HDFS client itself.

Previously the (single) processor pool closed down the client; this was moved into the server shutdown sequence. (Right now we have multiple processor pools, although we're going to try to get it back to a single one.)

Commit:
Remove some redundant test dependencies.

The main code already depends on these artefacts.

Commit:
Factor out versions where we have multiple artefacts that use the same version.

Commit:
Merge the test suites for testing various source and sink scenarios.

krisgeus commented:
Is there a need to change the MappingTestServer to incorporate the changes using multiple schema's? 
I think it can be used as is since you can specify a schema on the cmd line, but i think it would be useful if it takes the schema location(s) from the config

Commit:
Fix a bug where the schema registry threw an exception while handling shared sinks.

Commit:
Add some tests for various source/mapping/sink configurations.

Commit:
Add missing copyright header.

Commit:
Factor out most of the browser-source initialisation from the server class.

This allows source initialisation to take place in parallel, a win because it tends to be fairly slow.

Commit:
Update the example configuration to use the new format.

Somewhat embarrassingly, the mapping hadn't been converted to groovy yet. (The mapping hasn't been tested.)

Commit:
Merge branch 'master' into dev/many-to-many

friso commented on src/main/java/io/divolte/server/Server.java:
```diff
@@ -61,29 +66,90 @@ public Server(final ValidatedConfiguration vc) {
+        // Now instantiate all the sources. We do this in parallel because instantiation can be quite slow.
+        final ImmutableMap<String, BrowserSource> sources =
+                vc.configuration().sources.keySet()
+                  .parallelStream()
```
I'm not exactly sure what is implied by the parallel stream, other than that it is *possibly* parallel (according to API docs). Did anyone have a look into the implementation to figure out whether it's actually going to do instantiation in parallel?

asnare replied:
So, it does actually run it in parallel using it's own ForkJoin pool. I added this because the JavaScript minification happens here, and that takes several seconds. This was a simple attempt to get the tests to run faster. It had limited impact because the Closure Compiler already does as much as it can in parallel, but was still a win.

friso commented on src/main/java/io/divolte/server/config/DivolteConfiguration.java:
```diff
@@ -1,37 +1,131 @@
+        this.mappings = mappings.orElseGet(() -> defaultMappingConfigurations(this.sources.keySet(), this.sinks.keySet()));
+    }
+
+    public BrowserSourceConfiguration getBrowserSourceConfiguration(final String sourceName) {
```
I have a preference toward creating a `asBrowserSourceConfiguration()` method on `SourceConfiguration` (and onward `asXxxConfiguration()` for others to come), instead of doing both the presence check and the type check here. This would leave the check for existence of the source (using `Map#containsKey(...)`) to the client code and only performs the type check as part of the config class' logic. In different contexts the calling code might have more meaningful error messages for a missing key compared to the general case.

asnare replied:
I think the latter is cleaner: `getSourceConfiguration(…)` with the expected class as a parameter.

The purpose of these methods, and not using the map directly, is to enforce the nullability constraints. Basically it's a programmer error, not a user error, to request a source or sink that doesn't exist or has the wrong type. So the preconditions simply enforce this, and the static analysis tools can tell that the result is never `null`. Going directly to the map with `.get(…)` means null-checking the results.

friso commented on src/main/java/io/divolte/server/config/DivolteConfiguration.java:
```diff
@@ -1,37 +1,131 @@
+        return (BrowserSourceConfiguration)sourceConfiguration;
+    }
+
+    public MappingConfiguration getMappingConfiguration(final String mappingName) {
```
Similar to above, I think the check for existence of the key is the responsibility of the calling code (Using `Map#containsKey(...)`).

friso commented on src/main/java/io/divolte/server/config/MappingConfiguration.java:
```diff
@@ -0,0 +1,55 @@
+    @JsonCreator
+    MappingConfiguration(final Optional<String> schemaFile,
+                         final Optional<String> mappingScriptFile,
+                         @JsonProperty(required = true)
```
If the reference.conf has a empty collection for the property (meaning the specific conf would add entries), I think it's not possible to not have the property present. In that case the required=true wouldn't really add anything.

We could think about adding `@NotEmpty` to these fields, meaning the config would not validate with zero of either sources, mappings or sinks.

asnare replied:
The `reference.conf` doesn't supply anything at all for sources/mappings/sinks because of the way merging works: defaults are handled in code. So this annotation means that users have to supply sources/sinks when they declare a mapping. (Without it, Jackson will happily supply `null`.)

`@NotEmpty` is interesting. Although indeed such a mapping is pointless, I can imagine that it might be useful when setting up and configuring a server to have a mapping temporarily configured to have no sources or sinks. (It might also simplify template-based configuration.)

friso commented on src/main/java/io/divolte/server/kafka/KafkaFlusher.java:
```diff
@@ -81,8 +86,9 @@ public ProcessingDirective process(final Queue<AvroRecordBuffer> batch) {
+                         .map(i -> i.payload)
                          .map(this::buildRecord)
-                         .collect(Collectors.toList());
+                         .collect(Collectors.toCollection(() -> new ArrayList<>(batchSize)));
```
Much of the mechanics in here feel a bit allocation-heavy. I don't know the details of the approach, so cannot suggest an alternative right now, but perhaps something to look into.

asnare replied:
I looked into trying to reduce the number of collections we allocate and just pass streams around. However this wasn't possible:
 - We need to convert each event we're going to send into a Kafka message.
 - Each message we send to Kafka has its own `Future` to report the result.
 - On failure, we need to go back and correlate failures with the Kafka message, so we can preserve them for retry.

(It would be handy if we had (`.zip(…)`) to help with this. There might be a better way, but old-fashioned looping with an index works.)

So the best I could come up with for now was a list for:
 - The Kafka messages we're going to send.
 - The future responses for the messages in flight.
 - The failed messages, which need to be retried.

(The first two are pre-sized, the last isn't but Guava optimises for the common case.)

An alternative I considered was collecting the messages and corresponding futures into a single collection of tuples, albeit a custom container because we don't do tuples. However this is even more allocation-heavy as far as I can tell.

Something I didn't consider but would be slightly lighter-weight is native arrays instead of collections.

Commit:
Refactor the way source configurations are retrieved from the divolte configuration for consistency.

Retrieving source configurations is now the same as for sinks. In addition, we now universally throw IllegalArgumentException if the the wrong name and/or type is used. (This is a programming mistake, not a user error, as the javadoc now indicates.)

Commit:
Annotate some methods whose overrides should invoke super.

friso committed:
Move all mapping onto a single thread pool.

friso committed:
Resolve merge.

Commit:
Reflow some of the streaming.

Also substitute some method handles.

Commit:
Merge branch 'master' into dev/many-to-many

# Conflicts:
#	build.gradle

Commit:
Merge branch 'master' into dev/many-to-many

Commit:
Merge branch 'master' into dev/many-to-many

Commit:
Merge branch 'dev/version-bumps' into dev/many-to-many

Commit:
Merge branch 'master' into dev/many-to-many

Commit:
Consistent null annotation and default handling during configuration initialisation.

Commit:
Fix a typo in the .toString() implementation.

Commit:
Speed up tests by disabling HDFS support unless it's needed for the test.

Commit:
Merge branch 'master' into dev/many-to-many

Commit:
Refactor to use a builder for the list of errors.

Commit:
Return the value instead of assigning to intermediate reference.

Commit:
Remove unnecessary exceptions from throw specification.

Commit:
Remove redundant type bounds.

Commit:
Avoid Optional.get(); the intent here is that the reference is available, and if it's not that's exceptional.

Commit:
Merge branch 'master' into dev/many-to-many

Commit:
Merge branch 'master' into dev/many-to-many

Commit:
Merge branch 'master' into dev/many-to-many

Commit:
Merge branch 'master' into dev/many-to-many

Commit:
Merge branch 'master' into dev/many-to-many

Commit:
Merge branch 'master' into dev/many-to-many

Commit:
Merge branch 'master' into dev/many-to-many

Commit:
Update the configuration file we bundle with the RPM to use the new layout.

Commit:
Update the 'Getting Started' documentation to match the new configuration format.

Commit:
Ask sphinx to be picky about problems.

Commit:
Update link in comment to refer to the documentation for the specific version of Kafka that we're using.

Commit:
Introduce event-flows, and mention that we now support many-to-many mappings in Divolte.

Commit:
Bump documentation copyright.

Commit:
Update the configuration reference to describe the new way Divolte is configured.

Commit:
Update links to refer to the Avro specification version that we use.

Commit:
Capitalize list items.

Commit:
Render HTML as code block.

Commit:
Use Sphinx semantic markup where possible.

Commit:
Clarify that a specified schema is ignored by a mapping unless the script is specified.

Commit:
Update the mapping documentation.

Changes include Sphinx markup, and updates to reflect that we now support more than a single mapping.

Commit:
Sphinx markup for the example strings for user-agent components.

Plus a few typo corrections.

Commit:
Metro codes aren't an ISO thing.

Commit:
Latitude and longitude aren't really English.

Commit:
GeoNames hyperlink reference.

Commit:
Minor wording changes.

Commit:
Whitespace.

Commit:
Refresh the deployment documentation.

Commit:
Merge branch 'master' into dev/many-to-many

Commit:
Merge branch 'master' into dev/many-to-many

Commit:
Fix some failing Selenium tests.

Commit:
Call out the implicit default sources/mapping/sinks with a heading.

icychocolate98 commented:
Good morning,

I'm trying to send 2 custom events to kafka; the idea is send each event to its topic; but at the moment I receive both events in both topics.

I have configured 2 mappings with different schemas and 2 sinks (one for each mapping).

I don't know what is exactly happening, do you hace any idea?

Thanks !
Manu

soufianeodf commented:
Did you find any solution for that ? I'm also running in the same issue.
